File: examples/hello_service/hello_service_cpp/src/hello_server.cpp
/**
 *  A minimal ROS2 cpp node that creates a service.
 *
 *  Basic structure for ROS2 node:
 *  - Initialise rclcpp
 *  - Create a node
 *  - Create the service
 *  - Keep it alive with spin
 *  - Shutdown
 *
 *  RELIABLE and VOLATILE have been selected for the QOS_PROFILE. The reason with BEST_EFFORT,
 *  is that we won't be able to call the service from the terminal (CLI) and with TRANSIENT_LOCAL
 *  that can cause the service servers to receive requests from clients that have since terminated.
 *
 *  Author: Ioannis Selinis <selinis.g@gmail.com> 2025
 */

#include <rclcpp/rclcpp.hpp>
#include <example_interfaces/srv/add_two_ints.hpp>


static const rmw_qos_profile_t rmw_qos_profile =
{
    RMW_QOS_POLICY_HISTORY_KEEP_LAST,
    1,
    RMW_QOS_POLICY_RELIABILITY_RELIABLE,
    RMW_QOS_POLICY_DURABILITY_VOLATILE,
    RMW_QOS_DEADLINE_DEFAULT,
    RMW_QOS_LIFESPAN_DEFAULT,
    RMW_QOS_POLICY_LIVELINESS_SYSTEM_DEFAULT,
    RMW_QOS_LIVELINESS_LEASE_DURATION_DEFAULT,
    false
};


class HelloServer : public rclcpp::Node
{
    public:
        HelloServer() : Node("hello_c_server")
        {
            m_server = this->create_service<example_interfaces::srv::AddTwoInts>("add_two_ints",
                                                                                 std::bind(&HelloServer::serviceCallback, this, std::placeholders::_1, std::placeholders::_2),
                                                                                 rmw_qos_profile);
            RCLCPP_INFO(this->get_logger(), "Node has started");
        }

    private:
        void serviceCallback(const example_interfaces::srv::AddTwoInts::Request::SharedPtr request,
                             const example_interfaces::srv::AddTwoInts::Response::SharedPtr response)
        {
            response->sum = request->a + request->b;
            RCLCPP_INFO(this->get_logger(), "Request received with %d + %d = %d", (int) request->a, (int) request->b, (int) response->sum);
        }
        rclcpp::Service<example_interfaces::srv::AddTwoInts>::SharedPtr m_server;
};


int main (int argc, char **argv)
{
    rclcpp::init(argc, argv);                       // Initialise rclcpp (ROS communications)
    auto node = std::make_shared<HelloServer>();    // Smart/Shared pointer
    rclcpp::spin(node);                             // Keep node alive
    rclcpp::shutdown();                             // Shutdown executor
}

File: examples/hello_service/hello_service_py/hello_service_py/hello_server.py
#!/usr/bin/env python3

'''
    A minimal ROS2 python node that creates a service.

    Basic structure for ROS2 node publisher:
    - Initialise rclpy
    - Create a node
    - Create the server
    - Keep it alive with spin
    - Shutdown

    RELIABLE and VOLATILE have been selected for the QOS_PROFILE. The reason is with BEST_EFFORT
    is that we won't be able to call the service from the terminal (CLI) and with TRANSIENT_LOCAL
    that can cause the service servers to receive requests from clients that have since terminated.

    Author: Ioannis Selinis <selinis.g@gmail.com> 2025
'''

import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy, DurabilityPolicy
from example_interfaces.srv import AddTwoInts


# https://docs.ros.org/en/rolling/Concepts/Intermediate/About-Quality-of-Service-Settings.html
QOS_PROFILE = QoSProfile(
    reliability=ReliabilityPolicy.RELIABLE,         # RELIABLE or BEST_EFFORT
    durability=DurabilityPolicy.VOLATILE,           # VOLATILE or TRANSIENT_LOCAL
    history=HistoryPolicy.KEEP_LAST,                # KEEP_ALL or KEEP_LAST
    depth=1                                         # Used only with KEEP_LAST
)


class HelloServer(Node):
    """The HelloServer class that inherits from rclpy.node.Node.
    Initialises with server/service that returns the sum of two ints.
    """
    def __init__(self):
        super().__init__(node_name='hello_server')

        self.server = self.create_service(srv_type=AddTwoInts,
                                          srv_name='add_two_ints',
                                          callback=self.service_callback,
                                          qos_profile=QOS_PROFILE)


    def service_callback(self, request: AddTwoInts.Request, response: AddTwoInts.Response) -> AddTwoInts.Response:
        """The callback service to return the sum of two integers.

        Keyword arguments:
        request (AddTwoInts.Request) -- The request with the two integers
        response (AddTwoInts.Response) -- The response with the sum of the integers

        Return:
        AddTwoInts.Response -- The sum of the integers
        """
        response.sum = request.a + request.b
        self.get_logger().info(f'Request received with {request.a} + {request.b} = {response.sum}')
        return response


def main(args=None) -> None:
    """Entry Point for our node"""
    rclpy.init(args=args)   # Initialise rclpy (ROS communications)
    node = HelloServer()
    rclpy.spin(node=node)   # Keep node alive
    node.destroy_node()     # Destroy node
    rclpy.shutdown()        # Shutdown the executor


if __name__ == '__main__':
    main()
